src/hotspot/share/opto/graphKit.cpp

```diff
@@ -3969,20 +3971,28 @@ Node* GraphKit::new_array(Node* klass_node,     // array klass (maybe variable)
     initial_slow_test = initial_slow_test->as_Bool()->as_int_value(&_gvn);
   }
 
+  const TypeOopPtr* ary_type = _gvn.type(klass_node)->is_klassptr()->as_instance_type();
+  Node* valid_length_test = _gvn.intcon(1);
+  if (ary_type->klass()->is_array_klass()) {
+    BasicType bt = ary_type->klass()->as_array_klass()->element_type()->basic_type();
```

Comment:
FTR `array_element_basic_type()` is more appropriate here. 
`basic_type()` reports `T_INT` for all sub-word element types and it may lead to underestimation of maximum array length.

Reply:
Thinking more about it, it seems like a bug. For a large enough allocation, it may erroneously classify it as erroneous. Will it trigger wrong exception being thrown?

Reply:
You are right. (Was looking at it in the context of #6717.)

I missed that it is `ciArrayKlass::element_type()` which returns `ciType*` (instead of `Type*`) and `ciTypeArrayKlass::element_type()` relies on `Klass::layout_helper_element_type` to acquire proper element type. 

Sorry for the noise.